### PR TITLE
Fix invalid fusion of Matmul and Mul

### LINKED ISSUE #39572: The New Converter of Tensorflow 2.2.0 generate incorrect model output
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): windows/ubuntu
- TensorFlow installed from (source or binary): pip
- TensorFlow version (or github SHA if from source): 2.2.0

**Command used to run the converter or code if you’re using the Python API**

``` python
https://colab.research.google.com/drive/1uMiHGTFrj7R_rFcGJoje3Yfbap0sC9H-?usp=sharing
import tensorflow as tf
import numpy as np
np.set_printoptions(suppress=True)

length = 66

a = tf.constant(
    np.random.rand(length,length).astype(np.float32),
    shape=[length, length])

c = [ tf.constant(
    np.random.rand(length).astype(np.float32),
    shape=[ length]) for i in range(0,3)]


@tf.function
def func2(x ,c1, c2):
    return  tf.multiply(x, c2) + c2

@tf.function
def func(x):
    return  func2(tf.nn.bias_add(tf.matmul(x,a) , c[0] ), c[1],  c[2])
     


input =  np.random.rand(1,length).astype(np.float32)
original_output = func(input).numpy()

print("Orginal:")
print(original_output)


def save_and_test(experimental_new_converter , filename):

    lite = tf.lite.TFLiteConverter.from_concrete_functions([func.get_concrete_function(x = tf.TensorSpec(shape=[None,length], dtype=tf.float32) )])
    lite.experimental_new_converter  = experimental_new_converter
    open(filename,"wb").write(lite.convert())

    interpreter = tf.lite.Interpreter(model_path=filename)
    interpreter.allocate_tensors()
    interpreter.set_tensor(interpreter.get_input_details()[0]['index'], input)
    interpreter.invoke()

    print(filename, ":")
    output = interpreter.get_tensor(interpreter.get_output_details()[0]['index'])
    print(output)
    print(filename, " sum of absolute difference: ", np.sum(np.absolute(original_output - output)))


save_and_test(True, "mlir")
save_and_test(False, "toco")
```

**The output from the converter invocation**

``` 
Orginal:
[[ 6.9496317   7.9791102   0.23810546  1.6724037   9.687994   11.553456
   0.60509163  8.606714   17.731003    2.52809    16.075905    0.50299734
   6.119868   12.388639    8.383967   11.989329    7.3116794   8.7486725
   9.714089   12.450738    1.4212162   7.567941   14.836599    5.73806
   8.028838    2.5774102   7.9609547  19.35534    11.474097    1.2807347
   5.075203   17.249842    3.0866256  17.775652    7.390017   15.896519
  16.136719    6.5862875  12.216141    0.33137095 18.97356     1.9778614
   5.3641543   1.9284656   1.8500257  16.767817   12.7209     14.047361
   7.367909    3.1752422   5.949711   15.742522    2.2939441  16.137108
   1.6687201  13.069997    8.666047   16.782955    1.3117387   6.3798137
   6.557558   15.286772    7.5772853   9.168247    3.2267199   1.5568383 ]]

mlir :
[[ 8.160242   9.0523     7.3261642  8.867854   7.860682   8.722304
   7.2230916  8.622743   9.271859   7.7777925  9.489797   7.390472
   8.524878   9.906944   8.355099   8.330111   8.341421   8.271743
   8.937096   8.515015   7.225615   9.190289   8.9844675  8.162167
   8.694139   6.9362564  9.618863  10.1376915  8.462459   8.29322
   8.158456   8.796178   8.039083   9.774652   8.209907   9.378836
  10.289057   7.984726   8.897137   9.119129   9.802201   8.349303
   6.498592   8.638722   8.569397   9.730883   8.358421   9.018978
   8.4079275  6.496749   8.111706   8.541293   9.513434   9.01988
   7.2715945  8.307488   8.566431   9.748925   8.775068   7.9520493
   7.9917192  9.261906   7.9303417  9.471784   9.427535   7.840794 ]]
mlir  sum of absolute difference:  284.37515

toco :
[[ 6.9496307   7.9791093   0.23810542  1.6724037   9.687995   11.553459
   0.6050917   8.606714   17.731003    2.5280902  16.075907    0.50299734
   6.1198673  12.388639    8.383966   11.98933     7.311679    8.748673
   9.714088   12.450737    1.4212162   7.567941   14.836597    5.73806
   8.028838    2.5774097   7.960953   19.355337   11.474101    1.2807347
   5.075203   17.24984     3.086626   17.775652    7.390019   15.896517
  16.13672     6.586289   12.216139    0.33137095 18.973562    1.9778614
   5.364155    1.9284654   1.8500254  16.767818   12.7209015  14.047361
   7.367908    3.1752434   5.94971    15.74252     2.2939441  16.137106
   1.6687204  13.069999    8.666047   16.782953    1.3117385   6.379814
   6.557558   15.286772    7.577286    9.168246    3.2267199   1.5568382 ]]
toco  sum of absolute difference:  5.4582953e-05
```

**Failure details**
The conversion is successful, but the generated model is wrong,
The result running with the model is much different from the original result.

### PR DESCRIPTION
**Pull request by ghost**
This PR fix #39572, add some test cases.

### CODE REVIEW

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 1]
```diff
@@ -354,26 +354,26 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
       return failure();
     if (fc_op.fused_activation_function() != "NONE") return failure();
 
-    // Broadcast the constant operand of Mul if it isn't compatible to the
-    // filter input. We only support broadcasting the operand along the depth
-    // dimension, when the operand's depth is 1.
+    // Check if it is possible to fuse the constant mathematically
```
Comment (liufengdb) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:357]: can you add some comments on what are the conditions?

File: tensorflow/compiler/mlir/tensorflow/transforms/optimize.td [hunk 2]
```diff
@@ -17,8 +17,9 @@ include "mlir/Dialect/StandardOps/IR/Ops.td"
 include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td"
 
 def IsDataFormatNHWC : ConstantAttr<TF_ConvnetDataFormatAttr, "NHWC">;
-def BroadcastableElements :
-    Constraint<CPred<"TFL::IsBroadcastableElementsAttrs($0, $1)">>;
+def CanFuseMulAndConv2D :
+    Constraint<CPred<"TFL::IsBroadcastableElementsAttrs($0, $1) && TFL::CanFuseAffineOpThenMul($1)">>;
```
Comment (liufengdb) [inline tensorflow/compiler/mlir/tensorflow/transforms/optimize.td:21]: why mul->conv2d needs a condition of `affine->mul`? can you rename CanFuseAffineOpThenMul to a more general name?

**Review by liufengdb — COMMENTED** [review]
Looks great overall.

File: tensorflow/compiler/mlir/tensorflow/transforms/optimize.td [hunk 3]
```diff
@@ -17,8 +17,14 @@ include "mlir/Dialect/StandardOps/IR/Ops.td"
 include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td"
 
 def IsDataFormatNHWC : ConstantAttr<TF_ConvnetDataFormatAttr, "NHWC">;
-def BroadcastableElements :
-    Constraint<CPred<"TFL::IsBroadcastableElementsAttrs($0, $1)">>;
+
+//output[b, i, j, k] =
```
Comment (liufengdb) [inline tensorflow/compiler/mlir/tensorflow/transforms/optimize.td:21]: nit: one space after `//`

File: tensorflow/compiler/mlir/lite/utils/validators.cc [hunk 4]
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
...
 
+bool IsDimensionsDegenerateExceptLastOne(const ArrayRef<int64_t> elements_shape) {
+  for (auto i = 0; i < elements_shape.size() - 1; i++) {
+    if (elements_shape[i] != 1) return false;
+  }
+  return true;
+}
+
+bool IsDimensionsDegenerateExceptLastOne(Attribute val) {
+  if (const auto elements = val.dyn_cast<DenseElementsAttr>()) {
```
Comment (liufengdb) [inline tensorflow/compiler/mlir/lite/utils/validators.cc:89]: nit: remove `const` since mlir attribute by design is immutable.

**Review by liufengdb — DISMISSED** [review]

**Comment by gbaned** [comment]
@kroware Can you please address Ubuntu Sanity errors? Thanks!

**Comment by gbaned** [comment]
@liufengdb Can you please review this PR ? Thanks!

**Comment by lgeiger** [comment]
@liufengdb @joker-eph could you take another look at this? I seems like this PR fixes #39572 which sounds like a quite serious issue.

File: tensorflow/compiler/mlir/lite/utils/validators.h [hunk 5]
```diff
@@ -70,6 +70,8 @@ inline bool TFPaddingIsSameOrValid(Operation *op, StringAttr *padding) {
 /// Returns whether the given `a` and `b` have broadcast-compatible
 /// types.
 bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b);
+bool IsDimensionsDegenerateExceptLastOne(mlir::Attribute val);
+bool IsDimensionsDegenerateExceptLastOne(const ArrayRef<int64_t> elements_shape);
```
Comment (liufengdb) [inline tensorflow/compiler/mlir/lite/utils/validators.h:74]: nit, don't need the `const`

**Review by liufengdb — APPROVED** [review]

**Comment by gbaned** [comment]
@ghost  Can you please check @liufengdb's comments and keep us posted. Thanks!

**Comment by lgeiger** [comment]
> @ghost Can you please check @liufengdb's comments and keep us posted. Thanks!

@gbaned The GitHub user profile of the original PR author has been deleted and doesn't exist anymore, so I don't think the request will reach anyone. I am happy to address the small comment in a follow up PR after this has been merged as I am currently running into this issue myself and would really like to see this fix land.

**Comment by rthadur** [comment]
@lgeiger sure you can submit a new PR

**Comment by rthadur** [comment]
@lgeiger did you submit a new PR ? Closing this PR due to lack of actively , feel free to @mention us to reopen the PR.

**Comment by lgeiger** [comment]
@rthadur I didn't submit a new PR as I thought this would be merged soon. Will do so now.

**Pull request closed without merging**